File: strategy-ui/app/services/strategy_validate.py
from __future__ import annotations

from typing import Any, Dict, List

def validate_strategy(strategy: Dict[str, Any]) -> None:
    # Minimal structural checks
    for k in ("entry", "exits", "stops", "execution"):
        if k not in strategy:
            raise ValueError(f"strategy missing '{k}'")

    exits = strategy["exits"] or {}
    stops = strategy["stops"] or {}

    targets = exits.get("targets") or []
    if not isinstance(targets, list):
        raise ValueError("exits.targets must be a list")

    size_sum = 0.0
    for t in targets:
        sp = float(t.get("size_pct", 0))
        pp = float(t.get("profit_pct", 0))
        if sp <= 0 or sp > 100:
            raise ValueError("each target.size_pct must be in (0,100]")
        if pp <= 0:
            raise ValueError("each target.profit_pct must be > 0")
        size_sum += sp
    if size_sum > 100.0 + 1e-9:
        raise ValueError("targets size_pct sum must be <= 100")

    trailing = exits.get("trailing") or {}
    trailing_enabled = bool(trailing.get("enabled", False))
    if trailing_enabled:
        trail_pct = float(trailing.get("trail_pct", 0))
        act = float(trailing.get("activate_profit_pct", 0))
        if trail_pct <= 0:
            raise ValueError("trailing.trail_pct must be > 0 when enabled")
        if act < 0:
            raise ValueError("trailing.activate_profit_pct must be >= 0")

    time_exit = exits.get("time_exit") or {}
    time_enabled = bool(time_exit.get("enabled", False))
    if time_enabled:
        mx = int(time_exit.get("max_candles_in_trade", 0))
        if mx <= 0:
            raise ValueError("time_exit.max_candles_in_trade must be > 0 when enabled")

    stop_loss_pct = float(stops.get("stop_loss_pct", 0) or 0)
    if stop_loss_pct < 0:
        raise ValueError("stops.stop_loss_pct must be >= 0")

    # Require at least one exit path:
    has_exit = (len(targets) > 0) or trailing_enabled or time_enabled or (stop_loss_pct > 0)
    if not has_exit:
        raise ValueError("strategy must define at least one exit path (targets/trailing/time/stop)")

    execution = strategy["execution"] or {}
    fill_model = execution.get("fill_model")
    if fill_model not in ("open", "close"):
        raise ValueError("execution.fill_model must be 'open' or 'close'")

    fee_bps = float(execution.get("fee_bps", 0) or 0)
    slippage_bps = float(execution.get("slippage_bps", 0) or 0)
    if fee_bps < 0 or slippage_bps < 0:
        raise ValueError("fee_bps and slippage_bps must be >= 0")

File: packages/simulation/src/engine/strategy_validate.ts
/**
 * Strategy Validation
 *
 * Zod schemas for strategy JSON validation.
 * Matches Python strategy_validate.py logic.
 */

import { z } from 'zod';

const TargetSchema = z.object({
  size_pct: z.number().positive().max(100),
  profit_pct: z.number().positive(),
});

const TrailingConfigSchema = z.object({
  enabled: z.boolean().optional(),
  trail_pct: z.number().nonnegative().optional(),
  activate_profit_pct: z.number().nonnegative().optional(),
});

const TimeExitConfigSchema = z.object({
  enabled: z.boolean().optional(),
  max_candles_in_trade: z.number().int().positive().optional(),
});

const ExitsConfigSchema = z.object({
  targets: z.array(TargetSchema).optional(),
  trailing: TrailingConfigSchema.optional(),
  time_exit: TimeExitConfigSchema.optional(),
});

const StopsConfigSchema = z.object({
  stop_loss_pct: z.number().nonnegative(),
  break_even_after_first_target: z.boolean().optional(),
});

const EntrySignalSchema = z.object({
  type: z.enum(['rsi_below', 'ema_cross']),
  period: z.number().int().positive().optional(), // For RSI
  value: z.number().optional(), // For RSI threshold
  fast: z.number().int().positive().optional(), // For EMA cross
  slow: z.number().int().positive().optional(), // For EMA cross
  direction: z.enum(['bull', 'bear']).optional(), // For EMA cross
});

const EntryDelaySchema = z.object({
  mode: z.enum(['none', 'candles']),
  n: z.number().int().nonnegative().optional(),
});

const EntryConfigSchema = z.object({
  mode: z.enum(['immediate', 'signal']),
  signal: EntrySignalSchema.optional(),
  delay: EntryDelaySchema.optional(),
});

const ExecutionConfigSchema = z.object({
  fill_model: z.enum(['open', 'close']),
  fee_bps: z.number().nonnegative(),
  slippage_bps: z.number().nonnegative(),
});

export const StrategyConfigSchema = z.object({
  entry: EntryConfigSchema,
  exits: ExitsConfigSchema,
  stops: StopsConfigSchema,
  execution: ExecutionConfigSchema,
});

export type StrategyConfig = z.infer<typeof StrategyConfigSchema>;

/**
 * Validate strategy configuration
 *
 * @param strategy - Strategy configuration object
 * @throws Error if validation fails
 */
export function validateStrategy(strategy: unknown): asserts strategy is StrategyConfig {
  // Parse with Zod
  const result = StrategyConfigSchema.safeParse(strategy);
  if (!result.success) {
    const issues = result.error.issues.map((i) => `${i.path.join('.')}: ${i.message}`).join('; ');
    throw new Error(`Strategy validation failed: ${issues}`);
  }

  const config = result.data;

  // Check targets size sum <= 100
  const targets = config.exits?.targets || [];
  const sizeSum = targets.reduce((sum, t) => sum + t.size_pct, 0);
  if (sizeSum > 100.0 + 1e-9) {
    throw new Error('targets size_pct sum must be <= 100');
  }

  // Check trailing config if enabled
  const trailing = config.exits?.trailing;
  if (trailing?.enabled) {
    if (!trailing.trail_pct || trailing.trail_pct <= 0) {
      throw new Error('trailing.trail_pct must be > 0 when enabled');
    }
    if (trailing.activate_profit_pct !== undefined && trailing.activate_profit_pct < 0) {
      throw new Error('trailing.activate_profit_pct must be >= 0');
    }
  }

  // Check time exit config if enabled
  const timeExit = config.exits?.time_exit;
  if (timeExit?.enabled) {
    if (!timeExit.max_candles_in_trade || timeExit.max_candles_in_trade <= 0) {
      throw new Error('time_exit.max_candles_in_trade must be > 0 when enabled');
    }
  }

  // Require at least one exit path
  const hasExit =
    targets.length > 0 || trailing?.enabled || timeExit?.enabled || config.stops.stop_loss_pct > 0;

  if (!hasExit) {
    throw new Error('strategy must define at least one exit path (targets/trailing/time/stop)');
  }
}
